great_expectations/core/factory/checkpoint_factory.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from great_expectations._docs_decorators import public_api
+from great_expectations.checkpoint.checkpoint import Checkpoint
+from great_expectations.compatibility.typing_extensions import override
+from great_expectations.core.factory.factory import Factory
+from great_expectations.data_context.types.base import CheckpointConfig
+from great_expectations.exceptions import DataContextError
+
+if TYPE_CHECKING:
+    from great_expectations.data_context.data_context.abstract_data_context import (
+        AbstractDataContext,
+    )
+    from great_expectations.data_context.store import CheckpointStore
+    from great_expectations.data_context.types.resource_identifiers import (
+        ConfigurationIdentifier,
+        GXCloudIdentifier,
+    )
+
+
+# TODO: Add analytics as needed
+class CheckpointFactory(Factory[Checkpoint]):
+    def __init__(self, store: CheckpointStore, context: AbstractDataContext):
+        self._store = store
+        self._context = context
+
+    @public_api
+    @override
+    def add(self, checkpoint: Checkpoint) -> Checkpoint:
+        """Add a Checkpoint to the collection.
+
+        Parameters:
+            checkpoint: Checkpoint to add
+
+        Raises:
+            DataContextError if Checkpoint already exists
+        """
+        key = self._store.get_key(name=checkpoint.name, id=None)
+        if self._store.has_key(key=key):
+            raise DataContextError(
+                f"Cannot add Checkpoint with name {checkpoint.name} because it already exists."
+            )
+
+        self._store.add(key=key, value=checkpoint.get_config())
+
+        # TODO: Add id adding logic to CheckpointStore to prevent round trip
+        return self._get(key=key)
```

Comment:
Currently round tripping to get ID.

Ideally, we have the store add the ID to the input object but this is additional work. I can fast follow with a PR to update store logic but thought this was already a hefty PR.